feat(call-history): update the callhistory when the user reads missed calls

sokn-sys opened the pull request:
# COMPLETES # https://jira-eng-gpk2.cisco.com/jira/browse/SPARK-452240

## This pull request addresses

Created the new method to updateMissedCalls(), to update the missed calls. This method will post a request containing the session ID and end time of unread missed calls, along with the Webex access token included in the headers.​
After the post request, when the "**isRead**" item is added for a given user in the userSession response, Janus will trigger the **event:janus_user_viewed_session** mercury event to notify corresponding clients.


## by making the following changes
In the MSFT client side, when a user clicks on any call history record or the voicemail sub-tab, and an event is triggered and the userSession response is updated, the missed call notification badge/counter should dismiss.


https://github.com/webex/webex-js-sdk/assets/58249052/990c58aa-47ac-48db-8af4-85903b856681


<img width="966" alt="sdkTest" src="https://github.com/webex/webex-js-sdk/assets/58249052/ce317a60-a56e-411f-a4e7-2c20db47296b">
<img width="1723" alt="sdk1" src="https://github.com/webex/webex-js-sdk/assets/58249052/587e37b1-1a00-43f7-9d57-f6b3f54b99bc">
<img width="1728" alt="sdk2" src="https://github.com/webex/webex-js-sdk/assets/58249052/7bc76b2d-f541-46f3-b74b-2b16ed0e2b3a">
<img width="1728" alt="client3" src="https://github.com/webex/webex-js-sdk/assets/58249052/885532ff-a9db-48cd-bddd-3a432fedc953">

<img width="1725" alt="sdk_readStatus" src="https://github.com/webex/webex-js-sdk/assets/58249052/d435e71f-9bbe-4f91-9f3b-9a03bf3b191c">


### Change Type

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update
- [ ] Tooling change
- [ ] Internal code refactor

## The following scenarios where tested

< ENUMERATE TESTS PERFORMED, WHETHER MANUAL OR AUTOMATED >

### I certified that

- [x] I have read and followed [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request)
- [x] I discussed changes with code owners prior to submitting this pull request

- [x] I have not skipped any automated checks
- [x] All existing and new tests passed
- [x] I have updated the documentation accordingly

---

Make sure to have followed the [contributing guidelines](https://github.com/webex/webex-js-sdk/blob/master/CONTRIBUTING.md#submitting-a-pull-request) before submitting.

Shreyas281299 commented on packages/calling/src/CallHistory/types.ts:
```diff
@@ -41,4 +50,5 @@ export interface ICallHistory extends Eventing<CallHistoryEventTypes> {
     sort: SORT,
     sortBy: SORT_BY
   ): Promise<JanusResponseEvent>;
+  updateMissedCalls(endTimeSessionIds: EndTimeSessionId[]): Promise<UpdateMissedCallsResponse>;
```
Please add a description to the `updateMissedCalls` API, similar to getCallHistoryData.

Shreyas281299 reviewed (changes requested):
No big changes required. Code looks good.

sokn-sys commented on packages/calling/src/CallHistory/types.ts:
```diff
@@ -41,4 +50,5 @@ export interface ICallHistory extends Eventing<CallHistoryEventTypes> {
     sort: SORT,
     sortBy: SORT_BY
   ): Promise<JanusResponseEvent>;
+  updateMissedCalls(endTimeSessionIds: EndTimeSessionId[]): Promise<UpdateMissedCallsResponse>;
```
Done

sokn-sys commented on packages/calling/src/CallHistory/constants.ts:
```diff
@@ -1,9 +1,13 @@
 export const LIMIT = 50;
 export const NUMBER_OF_DAYS = 10;
 export const RESPONSE_MESSAGE = 'responseMessage';
+export const SET_READ_STATE = 'setReadState';
```
Done

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.test.ts:
```diff
@@ -115,5 +127,134 @@ describe('Call history tests', () => {
+    });
+  });
+
+  describe('updateMissedCalls success tests', () => {
```
Done

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.test.ts:
```diff
@@ -115,5 +127,134 @@ describe('Call history tests', () => {
       callSessionCallback(MOCK_SESSION_EVENT_LEGACY);
     });
+
+    it('verify the user viewed session event for missed calls update', (done) => {
```
Done

Shreyas281299 commented on packages/calling/src/CallHistory/CallHistory.test.ts:
```diff
@@ -115,5 +127,125 @@ describe('Call history tests', () => {
+        })
+      ) as jest.Mock;
+    });
+    it('successfully updates missed calls', async () => {
```
Leave a line after each test.

sokn-sys replied:
done

Kesari3008 commented on packages/calling/src/CallHistory/CallHistory.test.ts:
```diff
@@ -115,5 +127,129 @@ describe('Call history tests', () => {
+      }));
+      expect(response.statusCode).toEqual(200);
+      expect(response).toEqual(MOCK_UPDATE_MISSED_CALL_RESPONSE);
+      expect(global.fetch).toHaveBeenCalledTimes(1);
```
You can club these 2 expect statements into one: using toBeAclledOnceWith()

sokn-sys replied:
Done

Kesari3008 commented on packages/calling/src/CallHistory/CallHistory.test.ts:
```diff
@@ -115,5 +127,129 @@ describe('Call history tests', () => {
+      }));
+      expect(response).toStrictEqual(ERROR_DETAILS_401);
+      expect(response.statusCode).toBe(401);
+      expect(global.fetch).toHaveBeenCalledTimes(1);
```
same comment here as above

sokn-sys replied:
Done

Kesari3008 replied:
This is not addressed

Kesari3008 replied:
Also since the second argument here is the same object every time, instead of having all the fields to be copied again and again. Create an object to store it's value and pass that here

sokn-sys replied:
Done

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.ts:
```diff
@@ -127,12 +145,77 @@ export class CallHistory extends Eventing<CallHistoryEventTypes> implements ICal
+        throw new Error(`${response.status}`);
+      }
+
+      const data: UpdateMissedCallsResponse = await response.json();
```
The data field contains only the statusCode. We parse the response data and construct the response details.

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.ts:
```diff
@@ -127,12 +145,77 @@ export class CallHistory extends Eventing<CallHistoryEventTypes> implements ICal
   }
 
+  /**
+   * Function to update the missed call status in the call history using sessionId and time.
```
Done

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.ts:
```diff
@@ -145,6 +228,10 @@ export class CallHistory extends Eventing<CallHistoryEventTypes> implements ICal
     );
+    this.sdkConnector.registerListener<CallSessionViewedEvent>(
+      MOBIUS_EVENT_KEYS.CALL_SESSION_EVENT_VIEWED,
+      this.handleSessionEventsforReadData
```
Done

sokn-sys commented on packages/calling/src/CallHistory/CallHistory.ts:
```diff
@@ -127,12 +145,77 @@ export class CallHistory extends Eventing<CallHistoryEventTypes> implements ICal
+      const responseDetails: UpdateMissedCallsResponse = {
+        statusCode: data.statusCode as number,
+        data: {
+          data: SET_READ_STATE_SUCCESS_MESSAGE,
```
Done

Kesari3008 commented on packages/calling/src/CallHistory/types.ts:
```diff
@@ -41,4 +50,16 @@ export interface ICallHistory extends Eventing<CallHistoryEventTypes> {
     sortBy: SORT_BY
   ): Promise<JanusResponseEvent>;
+  /**
+   * This API `updateMissedCalls` is utilized to Updates the read state of user missed calls based on the specified parameters.
```
It should be update the read state not updates

sokn-sys replied:
Done

Kesari3008 reviewed (approved):
Approving with minor comments. Please take care of them before merging. Also did you test after changing the filed name from data to readStatusMessage? Ensure to test thoroughly before merge

sokn-sys commented:
> Approving with minor comments. Please take care of them before merging. Also did you test after changing the filed name from data to readStatusMessage? Ensure to test thoroughly before merge

 I've addressed the comments and tested thoroughly after changing the field name from "data" to "readStatusMessage." Everything is working as expected, and the tests are passing. We're good to merge
